### PR TITLE
Output csv's with the 500 closest words or refined words

### PR DESCRIPTION
**Pull request by Nancy078**
Output csv with the 500 closest words or refined words per dict/time period, with each word's cosine distance from that dict, that word's overall frequency, and a tuple including (a) each of the 3 closest terms in seed dictionary and (b) the cosine distance from that term to the focal term.

### CODE REVIEW

**Review by jhaber-zz — CHANGES_REQUESTED** [review]
Feedback: Great header and description at top, and nice job removing all print statements. To improve:

- what should be in 'output_dir' and 'work_dir'? Describe these and provide instructions
- include more headers, e.g. "import packages", "set filepaths", "define dictionaries"
- please include a comment with each code block describing what it's doing
- each function needs to have a docstring including a brief description and a list of parameters and things it returns
- each code block needs to have a description of what it's doing
- standardize spacing for clarity, e.g. `cult_exp1,dem_exp1 , relt_exp1=` --> `cult_exp1, dem_exp1, relt_exp1 = `

File: refine_dictionaries/get_cos_sim_dicts_df.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+cult_ref1, dem_ref1, relt_ref1=get_results_decade(m1, cult_core, dem_core, relt_core, cult_expanded_1, dem_expanded_1, relt_expanded_1)
+
+cult_ref2, dem_ref2, relt_ref2=get_results_decade(m2, cult_core, dem_core, relt_core, cult_expanded_2, dem_expanded_2, relt_expanded_2)
+
+cult_ref3, dem_ref3, relt_ref3=get_results_decade(m3, cult_core, dem_core, relt_core, cult_expanded_3, dem_expanded_3, relt_expanded_3)
+
+cult_ref4, dem_ref4, relt_ref4=get_results_decade(m4, cult_core, dem_core, relt_core, cult_expanded_4, dem_expanded_4, relt_expanded_4)
+
+
```
Comment (jhaber-zz) [inline refine_dictionaries/get_cos_sim_dicts_df.py:321]:
```suggestion
# Save similarities to disk in CSV format
```

File: refine_dictionaries/get_cos_sim_dicts_df.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+cult_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_cult_2004_2014.csv")
+dem_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_dem_2004_2014.csv")
+relt_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_relt_2004_2014.csv")
+
+
+####################################
+###### get similarities of refined dicts
+####################################
+
```
Comment (jhaber-zz) [inline refine_dictionaries/get_cos_sim_dicts_df.py:312]:
```suggestion

# run similarities function over all three dicts for each decade
```

File: refine_dictionaries/get_cos_sim_dicts_df.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            (500 terms closest core terms)   
+        - get_500 = boolean indicating whether to use generated 500 similar terms instead of using refined expanded terms
+   
+   outputs:
+       - cult_concat1 = cultural dataframe with the columns 'term', 'cosine_sim_with_core', ' closest_terms_from_core'
+       - dem_concat1 = demographic dataframe with the columns 'term', 'cosine_sim_with_core', ' closest_terms_from_core'
+       - relt_concat1 = relational dataframe with the columns 'term', 'cosine_sim_with_core', ' closest_terms_from_core'
+    
+    """
+
```
Comment (jhaber-zz) [inline refine_dictionaries/get_cos_sim_dicts_df.py:263]:
```suggestion
    
    # remove terms from each dictionary not in w2v model vocab
```

File: refine_dictionaries/get_cos_sim_dicts_df.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        - top_500_cult_1 = list of expanded terms for cultural perspective, 
+            could be empty list when getting similarities of expanded dicts 
+            (500 terms closest core terms)
+        - top_500_dem_1 = list of expanded terms for demographic perspective, 
+            could be empty list when getting similarities of expanded dicts 
+            (500 terms closest core terms)   
+        - top_500_relt_1 = list of expanded terms for relational perspective, 
+            could be empty list when getting similarities of expanded dicts 
+            (500 terms closest core terms)   
+        - get_500 = boolean indicating whether to use generated 500 similar terms instead of using refined expanded terms
```
Comment (jhaber-zz) [inline refine_dictionaries/get_cos_sim_dicts_df.py:255]: `get_500` should be a script input

File: refine_dictionaries/get_cos_sim_dicts_df.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+       - dem_concat1 = demographic dataframe with the columns 'term', 'cosine_sim_with_core', ' closest_terms_from_core'
+       - relt_concat1 = relational dataframe with the columns 'term', 'cosine_sim_with_core', ' closest_terms_from_core'
+    
+    """
+    
+    dem_core=remove_terms(dem_core, model)
+    cult_core=remove_terms(cult_core, model)
+    relt_core=remove_terms(relt_core, model)
+    
+    if get_500:
```
Comment (jhaber-zz) [inline refine_dictionaries/get_cos_sim_dicts_df.py:268]:
```suggestion
    # expand core dict to 500 closest terms if desired (get_500 command-line param)
    if get_500:
```

File: refine_dictionaries/get_cos_sim_dicts_df.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+all years w2v model
+- /models_storage/w2v_models/word2vec_ALLYEARS_phrased_filtered_enchant_orgdict_300d_10w_020423.bin
+
+core dictionaries
+cult_core = '/dictionary_methods/dictionaries/core/cultural_core.csv'
+dem_core = '/dictionary_methods/dictionaries/core/demographic_core.csv'
+relt_core = '/dictionary_methods/dictionaries/core/relational_core.csv'
+
+
+expanded decade dictionaries
```
Comment (jhaber-zz) [inline refine_dictionaries/get_cos_sim_dicts_df.py:47]:
```suggestion
expanded decade dictionaries (50 terms each)
```

**Review by jhaber-zz — CHANGES_REQUESTED** [review]
Excellent job incorporating my suggestions, from the comments to the clear variable names. Almost there! A few nitpicky things: 

- in function docstrings, the data type for each input and output should be indicated in parentheses. For instance, in `get_results_decade`, the input of `cult_core` should be listed as `cult_core (list of str)`, and `get_500` should `get_500 (bool)`.
-  I think there needs to be one more layer of comments within the commented blocks you provided here. I've provided a few specific suggestions to make this clear.
- It seems like you use the terms `core dict` and `refined dict` interchangeably, but this is inaccurate (and confusing). The core dicts have changed slightly, but they are still pegged at 36 terms each. I think `expanded` and `refined` mean close to the same thing, but I would stick to `refined` here when referring to the 50-term, hand-cleaned dicts unless you're talking about the 500-term expanded version (these are the closest words and not a true dict, but it's close enough). 
- I'm a bit confused as to the workflow here: Is it correct to say that "We start from the core dictionary (36 terms) and get the 500 closest words--or we just use the refined dictionary (50 terms)"--and get the similarities for each set to the core terms."?
- I think the parameter `get_500` in the `get_results_decade()` function needs to be a command-line input. Otherwise it needs to be manually changed within the script itself, making the script less portable and user-friendly. Please make that change and be sure to describe near the top what the parameter is for.

**Comment by jhaber-zz** [comment]
Also, please rename this script as `get_closest_words.py`, which I find a clearer title

File: refine_dictionaries/get_closest_words.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+from gensim.models import KeyedVectors
+import pandas as pd
+import sklearn
+import sklearn.metrics
+import numpy as np
+import os
+import argparse
+
+
+parser = argparse.ArgumentParser()
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:94]: ArgumentParser looks like a great tool for this. Strong execution, @Nancy078!

File: refine_dictionaries/get_closest_words.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+
+####################################
+###### load in decade specific models
+####################################
+
+m1 = KeyedVectors.load(work_dir+'/models_storage/w2v_models/word2vec_1971-1981_phrased_filtered_enchant_orgdict_300d_10w_020423.bin')
+m2 = KeyedVectors.load(work_dir+'/models_storage/w2v_models/word2vec_1982-1992_phrased_filtered_enchant_orgdict_300d_10w_020423.bin')
+m3 = KeyedVectors.load(work_dir+'/models_storage/w2v_models/word2vec_1993-2003_phrased_filtered_enchant_orgdict_300d_10w_020423.bin')
+m4 = KeyedVectors.load(work_dir+'/models_storage/w2v_models/word2vec_2004-2014_phrased_filtered_enchant_orgdict_300d_10w_020423.bin')
+ma =  KeyedVectors.load(work_dir+'/models_storage/w2v_models/word2vec_ALLYEARS_phrased_filtered_enchant_orgdict_300d_10w_020423.bin')
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:117]:
```suggestion
ma = KeyedVectors.load(work_dir+'/models_storage/w2v_models/word2vec_ALLYEARS_phrased_filtered_enchant_orgdict_300d_10w_020423.bin')
```

File: refine_dictionaries/get_closest_words.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+
+####################################
+###### load in expanded decade dictionaries
+####################################
+
+cult_refined_1 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/cultural_1971_1981.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+dem_refined_1 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/demographic_1971_1981.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+relt_refined_1 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/relational_1971_1981.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+
+
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:137]:
```suggestion
```

File: refine_dictionaries/get_closest_words.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+    relt_concat=get_output_df(relt_dict, relt_core,model)
+    return cult_concat, dem_concat, relt_concat
+
+
+
+####################################
+###### get similarities of expanded dicts (500)
+####################################
+
+def save_expanded_dicts_df():
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:297]: to make the code here more resilient, `output_dir` should be either passed in as a function argument or called via `global`. Even better would be to define the filepaths in the front end of the script, then use them as function parameters.

File: refine_dictionaries/get_closest_words.py [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+    dem_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_dem_2004_2014.csv")
+    relt_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_relt_2004_2014.csv")
+
+
+####################################
+###### get similarities of refined dicts
+####################################
+
+
+def save_refined_dicts_df():
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:333]: to make the code here more resilient, `output_dir` should be either passed in as a function argument or called via `global`. Even better would be to define the filepaths in the front end of the script, then use them as function parameters.

File: refine_dictionaries/get_closest_words.py [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+------------------------------------------------------------------------------------------------
+
+@description: Output csv with the 500 expanded terms (most cosine similar words from words in the core dictionaries) or refined terms (50 terms) per dict/time period, with each word's cosine distance from that dict, that word's overall frequency, and a tuple including (a) each of the 3 closest terms in seed dictionary and (b) the cosine distance from that term to the focal term.
+
+------------------------------------------------------------------------------------------------
+
+@script inputs: 
+- work_dir (-w) --> the current directory you're working with, it should be one level higher than models_storage and dictionary_methods
+- output_dir (-o) --> the directory where you want to store the csv dataframes
+- get_expanded_or_refined (-e) --> output csv's for expanded terms or refined terms
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:21]:
```suggestion
- get_expanded_or_refined (-e) --> output csvs with similar terms for expanded dictionaries (500 terms) or refined dictionaries (50 terms)
```

File: refine_dictionaries/get_closest_words.py [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+cult_refined_3 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/cultural_1993_2003.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+dem_refined_3 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/demographic_1993_2003.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+relt_refined_3 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/relational_1993_2003.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+
+cult_refined_4 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/cultural_2004_2014.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+dem_refined_4 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/demographic_2004_2014.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+relt_refined_4 = list(pd.read_csv(work_dir+'/dictionary_methods/dictionaries/expanded_decades/relational_2004_2014.txt', header=None)[0].apply(lambda x: x.replace(' ', '_')))
+
+####################################
+###### functions
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:151]:
```suggestion
###### define functions
```

File: refine_dictionaries/get_closest_words.py (hunk 10 above)
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:297]: Defining functions that execute the main functions and saves the output makes sense to me given you pass in the option of refining or expanding the dictionaries. However, the function names should make more clear what they are doing (not just saving), something more like below (drop the `df` part for simplicity). 

```suggestion
def get_results_save_expanded():
```

File: refine_dictionaries/get_closest_words.py (hunk 11 above)
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:333]: The function names should make more clear what they are doing (not just saving), something more like below (drop the `df` part for simplicity). 

```suggestion
def get_results_save_refined_dicts():
```

File: refine_dictionaries/get_closest_words.py [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+    cult_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_cult_2004_2014.csv")
+    dem_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_dem_2004_2014.csv")
+    relt_exp4.to_csv(output_dir+"/cos_sim_df_expanded_500_relt_2004_2014.csv")
+
+
+####################################
+###### get similarities of refined dicts
+####################################
+
+
```
Comment (jhaber-zz) [inline refine_dictionaries/get_closest_words.py:332]: No need for this comment as we are still defining functions. The function title and docstring should make clear what the function is for.
```suggestion
```

**Review by jhaber-zz — CHANGES_REQUESTED** [review]
@Nancy078 you have put a LOT of work into properly documenting these scripts, and the result here is an impressively clean, modular, and easy to follow script. Well done!

I have a few final requests for clean, reproducible coding here, most of which are simple name changes and the like. The most substantial request though is to define the filepaths in a way that requires only a single change to the script if and when we move it to a different folder/repo or use different file names (which we will soon). This will make modifying the script's contents later more straightforward. More instructions in my comments.
